tests/FormElement/CheckboxElementTest.php

```diff
@@ -106,4 +106,13 @@ public function testSetCheckValues()
             $checkbox
         );
     }
+
+    public function testCheckboxNotValidIfRequiredAndUnchecked()
+    {
+        $checkbox = new CheckboxElement('test');
+        $checkbox->setRequired();
+        $checkbox->setValue($checkbox->getUncheckedValue());
+
+        $this->assertFalse($checkbox->hasValue());
```

Comment:
It's correct that you patched `hasValue()`, but the original problem was that the checkbox was valid, hence this should be verified. (It's also what the test name states)

Setting the value directly is uncommon for checkboxes. The native way is to *check* them, or not. In this case, it is necessary to ensure it is not checked. Remember, unit tests should preferably be black box tests.

```suggestion
        $checkbox = new CheckboxElement('test');
        $checkbox->setRequired();
        $checkbox->setChecked(false);

        $this->assertFalse($checkbox->isValid());
```

Reply:
Yes, I agree with you. I had that before, but the test failed because of the `translate()` call on `null`.

$checkbox->isValid()
    $this->validate()
        $this->setMessages([$this->translate()])